src/main/java/org/commcare/core/network/ModernHttpRequester.java

```diff
@@ -41,149 +41,93 @@
-        } else if (!"https".equals(url.getProtocol())) {
-            throw new PlainTextPasswordException();
-        } else {
-            Authenticator.setDefault(new Authenticator() {
-                @Override
-                protected PasswordAuthentication getPasswordAuthentication() {
-                    return new PasswordAuthentication(username, password.toCharArray());
-                }
-            });
-        }
-    }
-
-    public void setResponseProcessor(HttpResponseProcessor responseProcessor) {
+        this.method = method;
+        this.requestBody = requestBody;
+        this.parts = parts;
+        this.commCareNetworkService = commCareNetworkService;
         this.responseProcessor = responseProcessor;
     }
 
-    public static class PlainTextPasswordException extends RuntimeException {
-    }
-
-    public void request() {
+    /**
+     * Executes and process the Request using the ResponseProcessor. Use makeRequest if you don't want ModernHTTPRequester to process the response.
+     */
+    @Nullable
+    public void processRequest() {
+        if (responseProcessor == null) {
+            throw new IllegalStateException("Please call makeRequest since responseProcessor is null");
+        }
         try {
-            httpConnection = setupConnection(buildUrl());
-            processResponse();
-        } catch (IOException e) {
+            response = makeRequest();
+            processResponse(responseProcessor, response.code(), this);
+        } catch (IOException | AuthenticationInterceptor.PlainTextPasswordException e) {
             e.printStackTrace();
-            responseProcessor.handleIOException(e);
-        } finally {
-            if (httpConnection != null) {
-                httpConnection.disconnect();
-            }
-        }
-    }
-
-    private URL buildUrl() throws IOException {
-        if (isPostRequest) {
-            return url;
-        } else {
-            return buildUrlWithParams();
+            responseProcessor.handleException(e);
         }
     }
 
-    protected HttpURLConnection setupConnection(URL builtUrl) throws IOException {
-        HttpURLConnection httpConnection = (HttpURLConnection)builtUrl.openConnection();
-        setupConnectionInner(httpConnection);
-        if (isPostRequest) {
-            httpConnection.setRequestMethod("POST");
-            httpConnection.setDoOutput(true);
-            buildPostPayload(httpConnection);
-        } else {
-            httpConnection.setRequestMethod("GET");
+    /**
+     * Executes the HTTP Request. Can be called directly to bypass response processor.
+     *
+     * @return Response from the HTTP call
+     * @throws IOException
+     */
+    public Response<ResponseBody> makeRequest() throws IOException {
+        switch (method) {
+            case POST:
+                currentCall = commCareNetworkService.makePostRequest(url, params, addPostHeaders(requestBody), requestBody);
```

Comment:
I was just trying to retain the earlier behaviour although current post headers are adding content Type and Content Length for the `POST` body and for a `MULTIPART_POST` this info would be different for different parts of the request. I think we can rename `addPostHeaders` as `addContentHeaders' to avoid the confusion.